src/main/java/org/opensearch/security/resources/ResourceAccessControlClient.java

```diff
@@ -0,0 +1,240 @@
+     * @param listener      Callback receiving a set of {@link ShareableResource} instances.
+     */
+    @Override
+    public <T extends ShareableResource> void listAllAccessibleResources(String resourceIndex, ActionListener<Set<T>> listener) {
+        if (handleIfFeatureDisabled("Unable to list all accessible resources.", listener)) return;
+
+        ListAccessibleResourcesRequest request = new ListAccessibleResourcesRequest.Builder().resourceIndex(resourceIndex).build();
+
+        resourceAccessHandler.getAccessibleResourcesForCurrentUser(resourceIndex, listener);
+    }
+
+    /**
+     * Checks whether security, resource sharing, or version compatibility is violated.
+     * If so, notifies the listener and returns {@code true}.
+     *
+     * @param contextMessage Context-specific message to append.
+     * @param listener       Listener to notify with an error.
+     * @return {@code true} if feature is disabled or unsupported; otherwise {@code false}.
+     */
+    private boolean handleIfFeatureDisabled(String contextMessage, ActionListener<?> listener) {
+        String reason = getFeatureDisabledReason();
+        if (!reason.isEmpty()) {
+            handleFeatureDisabled(reason + " " + contextMessage, listener);
+            return true;
+        }
+        return false;
+    }
+
+    /**
+     * Checks whether security, resource sharing, or version compatibility is violated.
+     * If so, responds with a default value and returns {@code true}.
+     *
+     * @param message         Context-specific log message.
+     * @param listener        Listener to notify with default response.
+     * @param defaultResponse Response to return when feature is disabled.
+     * @param <T>             Type of the default response.
+     * @return {@code true} if feature is disabled or unsupported; otherwise {@code false}.
+     */
+    private <T> boolean handleIfFeatureDisabled(String message, ActionListener<T> listener, T defaultResponse) {
+        String reason = getFeatureDisabledReason();
+        if (!reason.isEmpty()) {
+            log.debug("{} {}", reason, message);
+            listener.onResponse(defaultResponse);
+            return true;
+        }
+        return false;
+    }
+
+    /**
+     * Determines whether the security plugin, resource sharing, or version compatibility
+     * prevents the feature from being used.
+     *
+     * @return A non-empty message if the feature is disabled or unsupported; otherwise empty.
+     */
+    private String getFeatureDisabledReason() {
+        boolean sharingEnabled = settings.getAsBoolean(
+            ConfigConstants.OPENSEARCH_RESOURCE_SHARING_ENABLED,
+            ConfigConstants.OPENSEARCH_RESOURCE_SHARING_ENABLED_DEFAULT
+        );
+
+        if (!sharingEnabled) return "ShareableResource Access Control feature is disabled.";
+        Version nodeVersion = clusterService.localNode().getVersion();
+        if (nodeVersion.before(RESOURCE_SHARING_MIN_SUPPORTED_VERSION)) {
```

Comment:
Continuing discussion:https://github.com/opensearch-project/security/pull/5194#discussion_r2025237637 @nibix

Reply:
You are correct.  I'll remove this version control check.